File: Awful.apk/src/main/java/com/ferg/awfulapp/AwfulActivity.java
package com.ferg.awfulapp;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.graphics.Typeface;
import android.net.http.HttpResponseCache;
import android.os.Bundle;
import android.os.Message;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import com.androidquery.AQuery;
import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.network.NetworkUtils;
import com.ferg.awfulapp.preferences.AwfulPreferences;
import com.ferg.awfulapp.provider.ColorProvider;
import com.ferg.awfulapp.task.FeatureRequest;
import com.ferg.awfulapp.task.ProfileRequest;

import java.io.File;
import java.util.LinkedList;

/**
 * Convenience class to avoid having to call a configurator's lifecycle methods everywhere. This
 * class should avoid implementing things directly; the ActivityConfigurator does that job.
 * 
 * Most Activities in this awful app should extend this guy; that will provide things like locking
 * orientation according to user preference.
 * 
 * This class also provides a few helper methods for grabbing preferences and the like.
 */
public class AwfulActivity extends AppCompatActivity implements AwfulPreferences.AwfulPreferenceUpdate {
    protected static String TAG = "AwfulActivity";
    protected static final boolean DEBUG = Constants.DEBUG;
	private ActivityConfigurator mConf;
    
    private boolean loggedIn = false;
    
    protected AQuery aq;

    private TextView mTitleView;
    
    protected AwfulPreferences mPrefs;
    
    public void reauthenticate(){
    	NetworkUtils.clearLoginCookies(this);
        startActivityForResult(new Intent(this, AwfulLoginActivity.class), Constants.LOGIN_ACTIVITY_REQUEST);
    }
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        mPrefs = AwfulPreferences.getInstance(this,this);
        if(mPrefs.theme.equals(ColorProvider.DEFAULT) || mPrefs.theme.equals(ColorProvider.CLASSIC)){
            setTheme(R.style.Theme_AwfulTheme);
        }else if(mPrefs.theme.equals(ColorProvider.FYAD)){
            setTheme(R.style.Theme_AwfulTheme_FYAD);
        }else if(mPrefs.theme.equals(ColorProvider.BYOB)){
            setTheme(R.style.Theme_AwfulTheme_BYOB);
        }else if(mPrefs.theme.equals(ColorProvider.YOSPOS)){
            setTheme(R.style.Theme_AwfulTheme_YOSPOS);
        }else if(mPrefs.theme.equals(ColorProvider.AMBERPOS)){
            setTheme(R.style.Theme_AwfulTheme_AMBERPOS);
        }else{
            setTheme(R.style.Theme_AwfulTheme_Dark);
        }
        super.onCreate(savedInstanceState); if(DEBUG) Log.e(TAG, "onCreate");
        aq = new AQuery(this);
        mConf = new ActivityConfigurator(this);
        mConf.onCreate();
    }

    @Override
    protected void onStart() {
        super.onStart(); if(DEBUG) Log.e(TAG, "onStart");
        mConf.onStart();
    }
    
    @Override
    protected void onResume() {
        super.onResume(); if(DEBUG) Log.e(TAG, "onResume");
        mConf.onResume();
        // check login state when coming back into use, instead of in onCreate
        loggedIn = NetworkUtils.restoreLoginCookies(this.getAwfulApplication());
        if (isLoggedIn()) {
        	if(mPrefs.ignoreFormkey == null || mPrefs.userTitle == null){
        		 NetworkUtils.queueRequest(new ProfileRequest(this).build(null, null));
        	}
            if(mPrefs.ignoreFormkey == null){
                NetworkUtils.queueRequest(new FeatureRequest(this).build(null, null));
            }
            Log.v(TAG, "Cookie Loaded!");
        } else {
        	if(!(this instanceof AwfulLoginActivity)){
        		reauthenticate();
        	}
        }
    }
    
    @Override
    protected void onPause() {
        super.onPause(); if(DEBUG) Log.e(TAG, "onPause");
        mConf.onPause();
    }

    @SuppressLint("NewApi")
    @Override
    protected void onStop() {
        super.onStop(); if(DEBUG) Log.e(TAG, "onStop");
        mConf.onStop();
        HttpResponseCache cache = HttpResponseCache.getInstalled();
        if(cache != null){
            cache.flush();
        }
    }
    
    @Override
    protected void onDestroy() {
        super.onDestroy(); if(DEBUG) Log.e(TAG, "onDestroy");
        mConf.onDestroy();
        mPrefs.unregisterCallback(this);
    }


    @Override
	protected void onActivityResult(int request, int result, Intent intent) {
		super.onActivityResult(request, result, intent);
    	Log.w(TAG,"onActivityResult: " + request+" result: "+result);
		if(request == Constants.LOGIN_ACTIVITY_REQUEST && result == Activity.RESULT_CANCELED){
			finish();
		}
	}

	protected void setActionBar() {

        ActionBar action = getSupportActionBar();
        action.setDisplayShowTitleEnabled(false);
        action.setCustomView(R.layout.actionbar_title);
        mTitleView = (TextView) action.getCustomView();
        mTitleView.setMovementMethod(new ScrollingMovementMethod());
        updateActionbarTheme();
        action.setDisplayShowCustomEnabled(true);
        action.setDisplayHomeAsUpEnabled(true);
    }
    
    protected void updateActionbarTheme(){
        ActionBar action = getSupportActionBar();
        if(action != null && mTitleView != null){
	        //action.setBackgroundDrawable(new ColorDrawable(ColorProvider.getActionbarColor()));
	        mTitleView.setTextColor(ColorProvider.getActionbarFontColor());
	        setPreferredFont(mTitleView, Typeface.NORMAL);
        }
    }

    public void displayUserCP() {
    	displayForum(Constants.USERCP_ID, 1);
    }
    
    public void displayThread(int id, int page, int forumId, int forumPage, boolean forceReload){
    	startActivity(new Intent().setClass(this, ForumsIndexActivity.class)
    							  .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
    							  .putExtra(Constants.THREAD_ID, id)
    							  .putExtra(Constants.THREAD_PAGE, page)
    							  .putExtra(Constants.FORUM_ID, forumId)
    							  .putExtra(Constants.FORUM_PAGE, forumPage));
    }
    
    public void displayForum(int id, int page){
    	startActivity(new Intent().setClass(this, ForumsIndexActivity.class)
    							  .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
    							  .putExtra(Constants.FORUM_ID, id)
    							  .putExtra(Constants.FORUM_PAGE, page));
    }
    
    public void displayForumIndex(){
    	startActivity(new Intent().setClass(this, ForumsIndexActivity.class)
                .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP));
    }

    public void setActionbarTitle(String aTitle, Object requestor) {
        ActionBar action = getSupportActionBar();
        if(action != null){
        	mTitleView = (TextView) action.getCustomView();
        }
    	if(aTitle != null && mTitleView != null && aTitle.length()>0){
//    		mTitleView.setText(Html.fromHtml(aTitle));
    		mTitleView.setText(aTitle);
			mTitleView.scrollTo(0, 0);
    	}else{
    		Log.e(TAG, "FAILED setActionbarTitle - "+aTitle);
    	}
    }
    
    public AwfulApplication getAwfulApplication(){
    	return (AwfulApplication) getApplication();
    }
    
    public void setPreferredFont(View view, int flags){
    	if(getApplication() != null && view != null){
    		((AwfulApplication)getApplication()).setPreferredFont(view, flags);
    	}
    }
    
    public void setPreferredFont(View view){
    	setPreferredFont(view, -1);
    }

	@Override
	public void onPreferenceChange(AwfulPreferences prefs, String key) {
        Log.d(TAG, "Key changed: "+key);
        if("theme".equals(key) || "page_layout".equals(key)) {
            if (mPrefs.theme.equals(ColorProvider.DEFAULT) || mPrefs.theme.equals(ColorProvider.CLASSIC)) {
                setTheme(R.style.Theme_AwfulTheme);
            } else if (mPrefs.theme.equals(ColorProvider.FYAD)) {
                setTheme(R.style.Theme_AwfulTheme_FYAD);
            } else if (mPrefs.theme.equals(ColorProvider.BYOB)) {
                setTheme(R.style.Theme_AwfulTheme_BYOB);
            } else if (mPrefs.theme.equals(ColorProvider.YOSPOS)) {
                setTheme(R.style.Theme_AwfulTheme_YOSPOS);
            } else if (mPrefs.theme.equals(ColorProvider.AMBERPOS)) {
                setTheme(R.style.Theme_AwfulTheme_AMBERPOS);
            } else {
                setTheme(R.style.Theme_AwfulTheme_Dark);
            }
            afterThemeChange();
        }
		updateActionbarTheme();
	}
	
	protected boolean isLoggedIn(){
		if(!loggedIn){
			loggedIn = NetworkUtils.restoreLoginCookies(this.getAwfulApplication());
		}
		return loggedIn;
	}

	public boolean isFragmentVisible(AwfulFragment awfulFragment) {
		return true;
	}


	
    public void afterThemeChange() {
        recreate();
    }

	@Override
	public File getCacheDir() {
		Log.e(TAG,"getCacheDir(): "+super.getCacheDir());
		return super.getCacheDir();
	}
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/AwfulActivity.kt
package com.ferg.awfulapp

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.content.pm.ActivityInfo.*
import android.graphics.Typeface
import android.net.Uri
import android.net.http.HttpResponseCache
import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.v7.app.AppCompatActivity
import android.text.method.ScrollingMovementMethod
import android.view.View
import android.widget.TextView
import com.ferg.awfulapp.announcements.AnnouncementsManager
import com.ferg.awfulapp.constants.Constants
import com.ferg.awfulapp.constants.Constants.LOGIN_ACTIVITY_REQUEST
import com.ferg.awfulapp.network.NetworkUtils
import com.ferg.awfulapp.preferences.AwfulPreferences
import com.ferg.awfulapp.provider.AwfulTheme
import com.ferg.awfulapp.provider.ColorProvider
import com.ferg.awfulapp.task.AwfulRequest
import com.ferg.awfulapp.task.FeatureRequest
import com.ferg.awfulapp.task.ProfileRequest
import timber.log.Timber
import java.io.File

/**
 * Created by baka kaba on 18/01/2018.
 *
 * Base Activity class handling theming, orientation changes and access to preferences and callbacks.
 *
 * This is an update of the original AwfulActivity.java, but reworked to simplify the logic and
 * to pull in some code from other components, so it's more focused on how the app uses activities.
 *
 * All activities should inherit from this one, so you get free theming, preference handling and
 * access to app navigation. Call [setUpActionBar] after setting up your Toolbar to apply the standard
 * app style and behaviour.
 */
abstract class AwfulActivity : AppCompatActivity(), AwfulPreferences.AwfulPreferenceUpdate {
    private var customActivityTitle: TextView? = null
    // TODO: this is a var but honestly why does any activity need to replace it with their own copy? It's a singleton - if they're doing it for the callbacks, just use the register method
    var mPrefs: AwfulPreferences = AwfulPreferences.getInstance()


    //
    // Lifecycle
    //


    @CallSuper
    override fun onCreate(savedInstanceState: Bundle?) {
        Timber.i("*** onCreate")
        mPrefs.registerCallback(this)
        updateTheme()
        super.onCreate(savedInstanceState)
    }

    @CallSuper
    override fun onStart() {
        Timber.i("*** onStart")
        super.onStart()
    }

    @CallSuper
    override fun onResume() {
        Timber.i("*** onResume")
        super.onResume()
        updateOrientation()
        when {
            Authentication.isUserLoggedIn() -> with(mPrefs) {
                if (ignoreFormkey == null || userTitle == null) ProfileRequest(this@AwfulActivity).sendBlind()
                if (ignoreFormkey == null) FeatureRequest(this@AwfulActivity).sendBlind()
            }
        // TODO: this interferes with the code in AwfulFragment#reAuthenticate - i.e. it forces "return to this activity" behaviour, but fragments may want to return to the main activity.
        // And the activity isn't always the authority, e.g. using BasicActivity which is a plain container where all the specific behaviour is handled by its fragment. It's awkward
            this !is AwfulLoginActivity -> showLogIn(returnToMainActivity = false)
        }
    }

    private fun AwfulRequest<*>.sendBlind() = NetworkUtils.queueRequest(this.build())


    @CallSuper
    override fun onPause() {
        Timber.i("*** onPause")
        super.onPause()
    }

    @CallSuper
    @SuppressLint("NewApi")
    override fun onStop() {
        Timber.i("*** onStop")
        super.onStop()
        HttpResponseCache.getInstalled()?.flush()
    }


    @CallSuper
    override fun onDestroy() {
        Timber.i("*** onDestroy")
        super.onDestroy()
        mPrefs.unregisterCallback(this)
    }


    @CallSuper
    override fun onActivityResult(request: Int, result: Int, intent: Intent?) {
        Timber.i("onActivityResult: $request result: $result")
        super.onActivityResult(request, result, intent)
        supportFragmentManager.fragments.forEach { it.onActivityResult(request, result, intent) }
        if (request == LOGIN_ACTIVITY_REQUEST && result == Activity.RESULT_CANCELED) {
            Timber.w("Result from login activity - cancelled, closing app")
            finish()
        }
    }


    //
    // Action bar
    //

    /**
     * Initialises the Action Bar with the app's custom settings, behaviour and theming.
     */
    protected fun setUpActionBar() {
        // TODO: check if this plays nicely with activities without the custom title view, and make it optional if not - having a general "theme the action bar" method would be a good idea
        supportActionBar?.apply {
            setDisplayShowTitleEnabled(false)
            setCustomView(R.layout.actionbar_title)
            customActivityTitle = customView as TextView
            customActivityTitle?.movementMethod = ScrollingMovementMethod()
            updateActionbarTheme()
            setDisplayShowCustomEnabled(true)
            setDisplayHomeAsUpEnabled(true)
        }
    }


    /**
     * Set a (non-empty) title for the custom action bar.
     */
    open fun setActionbarTitle(title: String) {
        Timber.d("Setting action bar title: %s", title)
        supportActionBar?.apply { customActivityTitle = customView as TextView }
        with(customActivityTitle) {
            if (this == null || title.isEmpty()) {
                Timber.w("FAILED setActionbarTitle - $title")
            } else {
                text = title
                scrollTo(0, 0)
            }
        }
    }


    private fun updateActionbarTheme() {
        supportActionBar?.apply {
            customActivityTitle?.let { title ->
                title.setTextColor(ColorProvider.ACTION_BAR_TEXT.color)
                setPreferredFont(title, Typeface.NORMAL)
            }
        }
    }


    //
    // Preferences and other UI
    //

    override fun onPreferenceChange(prefs: AwfulPreferences, key: String?) {
        Timber.i("Key changed: ${key!!}")
        updateOrientation()
        if ("theme" == key || "page_layout" == key) {
            updateTheme()
            afterThemeChange()
        }
        updateActionbarTheme()
    }


    private fun updateOrientation() {
        requestedOrientation = when (mPrefs.orientation) {
            "portrait" -> SCREEN_ORIENTATION_PORTRAIT
            "landscape" -> SCREEN_ORIENTATION_LANDSCAPE
            "sensor" -> SCREEN_ORIENTATION_SENSOR
            else -> SCREEN_ORIENTATION_UNSPECIFIED
        }
    }

    @JvmOverloads
    fun setPreferredFont(view: View?, flags: Int = -1) =
            view?.let { (application as AwfulApplication).setPreferredFont(view, flags) }

    protected fun updateTheme() = setTheme(AwfulTheme.forForum(null).themeResId)

    // TODO: see if this can be rolled into updateTheme without causing issues
    private fun afterThemeChange() = recreate()


    //
    // Misc
    //

    override fun getCacheDir(): File {
        Timber.i("getCacheDir(): ${super.getCacheDir()}")
        return super.getCacheDir()
    }


    //
    // App navigation
    //

    open fun navigate(event: NavigationEvent) = startActivity(event.getIntent(applicationContext))

    /**
     * Log out, and show the login activity.
     *
     * If [returnToMainActivity] is true, after logging in the user will be directed to the main
     * activity, which receives the login activity result, otherwise the user returns to the current
     * activity. This is useful if e.g. the user jumps straight into their private messages, receives
     * a logged out error, and logs in - you can either update the private messages view, or kick them
     * to the main activity.
     */
    fun showLogIn(returnToMainActivity: Boolean = false) {
        if (returnToMainActivity) {
            startActivity(NavigationEvent.ReAuthenticate.getIntent(applicationContext))
        } else {
            Authentication.reAuthenticate(this)
        }
    }

    fun showLogoutDialog() = LogOutDialog().show(supportFragmentManager, "logout dialog")

    /** Display the announcements  */
    fun showAnnouncements() = AnnouncementsManager.getInstance().showAnnouncements(this)

    /**
     * Display the user's PMs
     *
     * @param openMessageUri a Uri to a private message you want to open, parsed from a new PM link on the site
     */
    fun showPrivateMessages(openMessageUri: Uri? = null) =
    // TODO: rework this so the message ID is parsed from the link, and we just pass that around internally
            Intent().apply {
                setClass(this@AwfulActivity, PrivateMessageActivity::class.java)
                openMessageUri?.let(::setData)
            }.let(::startActivity)

    /**
     * Display the post/reply/edit composer.
     *
     * @param threadId the ID of the thread the post is in
     * @param postType a post/reply/edit constant, see types in AwfulMessage
     * @param sourcePostId if we're editing or quoting a post, this is its ID
     */
    fun showPostComposer(threadId: Int, postType: Int, sourcePostId: Int) {
        // TODO: this should probably all be refactored into types like the NavigationEvents (maybe even rolled in with them) - discrete Posting events with the specific associated data for each
        startActivityForResult(
                Intent(this, PostReplyActivity::class.java)
                        .putExtra(Constants.REPLY_THREAD_ID, threadId)
                        .putExtra(Constants.EDITING, postType)
                        .putExtra(Constants.REPLY_POST_ID, sourcePostId),
                PostReplyFragment.REQUEST_POST
        )
    }


    companion object {
        val DEBUG = Constants.DEBUG
    }
}

